File: python/exer07_exec_service_main.py
'''
EXECUTOR SERVICE & THREAD POOL IMPLEMENTATION
'''

import time
from exer07_exec_service_itask import ITask
from exer07_exec_service_v0a import MyExecServiceV0A
from exer07_exec_service_v0b import MyExecServiceV0B
from exer07_exec_service_v1a import MyExecServiceV1A
from exer07_exec_service_v1b import MyExecServiceV1B
from exer07_exec_service_v2a import MyExecServiceV2A
from exer07_exec_service_v2b import MyExecServiceV2B



class MyTask(ITask):
    def __init__(self, task_id: str):
        self.id = task_id

    def run(self):
        print(f'Task {self.id} is starting')
        time.sleep(3)
        print(f'Task {self.id} is completed')



NUM_THREADS = 2
NUM_TASKS = 5

exec_service = MyExecServiceV0A(NUM_THREADS)

lsttask = [MyTask(chr(i + 65)) for i in range(NUM_TASKS)]

for task in lsttask:
    exec_service.submit(task)

print('All tasks are submitted')

exec_service.wait_task_done()
print('All tasks are completed')

exec_service.shutdown()

File: cpp-std/exer07-exec-service-main.cpp
/*
EXECUTOR SERVICE & THREAD POOL IMPLEMENTATION
*/


#include <iostream>
#include <chrono>
#include "exer07-exec-service-itask.hpp"
// #include "exer07-exec-service-v1.hpp"
// #include "exer07-exec-service-v2a.hpp"
#include "exer07-exec-service-v2b.hpp"



class MyTask : public ITask {
public:
    char id;

public:
    void run() override {
        std::cout << "Task " << id << " is starting" << std::endl;
        std::this_thread::sleep_for(std::chrono::seconds(3));
        std::cout << "Task " << id << " is completed" << std::endl;
    }
};



int main() {
    constexpr int NUM_THREADS = 2;
    constexpr int NUM_TASKS = 5;


    auto execService = MyExecServiceV2B();
    execService.init(NUM_THREADS);


    auto lstTask = std::vector<MyTask>(NUM_TASKS);

    for (int i = 0; i < NUM_TASKS; ++i)
        lstTask[i].id = 'A' + i;


    for (auto&& task : lstTask)
        execService.submit(&task);

    std::cout << "All tasks are submitted" << std::endl;


    execService.waitTaskDone();
    std::cout << "All tasks are completed" << std::endl;


    execService.shutdown();


    return 0;
}
